torch/distributions/wishart.py

```diff
@@ -0,0 +1,198 @@
+    assert x > (p - 1) / 2, "Wrong domain for digamma function."
+    return torch.digamma(x - torch.arange(p) / 2).sum()
+
+class Wishart(Distribution):
+    r"""
+    Creates a Wishart distribution parameterized by a square root matrix of symmetric matrix.
+    The Wishart distribution can be parameterized either in terms of
+    an outer product of general square root matrix e,g.,
+    :math:`\mathbf{\Sigma} = \mathbf{P}\mathbf{D}\mathbf{P}^\top = \mathbf{P'}\mathbf{P'}^\top` or
+    an outer product of cholesky decomposition :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top`
+    can be obtained via Cholesky decomposition of the covariance.
+    Example:
+        >>> m = Wishart(torch.eye(2), torch.Tensor([2]))
+        >>> m.sample()  #Wishart distributed with mean=`df * I` and
+                        #variance(x_ij)=`df` for i != j and variance(x_ij)=`2 * df` for i == j
+    Args:
+        covariance_matrix (Tensor): positive-definite covariance matrix
+        precision_matrix (Tensor): positive-definite precision matrix
+        scale_tril (Tensor): lower-triangular factor of covariance, with positive-valued diagonal
+        df (Tensor): real-valued parameter > larger than dimension of Square matrix
+    Note:
+        Only one of :attr:`covariance_matrix` or :attr:`precision_matrix` or
+        :attr:`scale_tril` can be specified.
+        Using :attr:`scale_tril` will be more efficient: all computations internally
+        are based on :attr:`scale_tril`. If :attr:`covariance_matrix` or
+        :attr:`precision_matrix` is passed instead, it is only used to compute
+        the corresponding lower triangular matrices using a Cholesky decomposition.
+        'torch.distributions.LKJCholesky' is a restricted Wishart distribution.[1]
+
+    **References**
+
+    [1] `On equivalence of the LKJ distribution and the restricted Wishart distribution`,
+    Zhenxun Wang, Yunan Wu, Haitao Chu.
+    """
+    arg_constraints = {
+        'covariance_matrix': constraints.positive_definite,
+        'precision_matrix': constraints.positive_definite,
+        'scale_tril': constraints.positive_definite,
+        'df': None,
+    }
+    support = constraints.positive_definite
+    has_rsample = True
+
+    def __init__(self, covariance_matrix=None, precision_matrix=None, scale_tril=None, df=None,
+                 bartlett_decomposition=True, validate_args=None):
+
+        assert (covariance_matrix is not None) + (scale_tril is not None) + (precision_matrix is not None) == 1, \
+            "Exactly one of covariance_matrix or precision_matrix or scale_tril may be specified."
+
+        if scale_tril is not None:
+            assert scale_tril.dim() > 1, \
+                "scale_tril matrix must be at least two-dimensional, with optional leading batch dimensions"
+            batch_shape = scale_tril.shape[:-2]
+            event_shape = scale_tril.shape[-2:]
+            self.scale_tril = scale_tril.expand(batch_shape + (-1, -1))
```

Comment:
Thank you for the delicate feedbacks. I didn't noticed that this part is redundant.
I followed the implementation practice in [MultivariateNormal](https://github.com/pytorch/pytorch/blob/a31aea8eaa99a5ff72b5d002c206cd68d5467a5e/torch/distributions/multivariate_normal.py#L78): [location](https://github.com/pytorch/pytorch/blob/a31aea8eaa99a5ff72b5d002c206cd68d5467a5e/torch/distributions/multivariate_normal.py#L130)
Do you want PR of this part, too?

Reply:
It should be fine there. The reason why we need that is because `batch_shape` is informed by both the shape of `loc` and `scale_tril`. Here on the other hand, we don't need to do this broadcasting op between two tensors, so `batch_shape` only depends on the shape of `scale_tril`.

Reply:
I am wondering if the we should consider multiple 'df' parameters and multiple covariance matrices when the input is given like that, like in some other distributions like gamma. Can you give me your advice?

Reply:
I suppose it depends on how common the use case of having batched `df` is. If we go by LKJCholesky, we haven't had anyone request a batched `df` feature. Unless it is relatively simple to implement batched `df`, I would suggest keeping it simple and only accept float values. If in the future, there is a request to batch this, we can have a more involved implementation.

Reply:
Thanks for your comment. It makes sense. I will keep the simplest version of  implementation now.